crates/vm/backends/levm.rs

```diff
@@ -1,376 +1,549 @@
+pub struct LevmGetStateTransitionsIn<'a> {
+    initial_state: &'a EvmState,
+    block_hash: H256,
+    new_state: &'a CacheDB,
+}
+
+impl<'a> LevmGetStateTransitionsIn<'a> {
+    pub fn new(initial_state: &'a EvmState, block_hash: H256, new_state: &'a CacheDB) -> Self {
+        LevmGetStateTransitionsIn {
+            initial_state,
+            block_hash,
+            new_state,
         }
     }
+}
 
-    // Account updates are initialized like this because of the beacon_root_contract_call, it is going to be empty if it wasn't called.
-    let mut account_updates = crate::get_state_transitions(state);
+/// Input for [LEVM::process_withdrawals]
+pub struct LevmProcessWithdrawalsIn<'a> {
+    block_cache: &'a mut CacheDB,
+    withdrawals: &'a [Withdrawal],
+    store: Option<&'a Store>,
+    parent_hash: H256,
+}
 
-    let mut receipts = Vec::new();
-    let mut cumulative_gas_used = 0;
+impl<'a> LevmProcessWithdrawalsIn<'a> {
+    pub fn new(
+        block_cache: &'a mut CacheDB,
+        withdrawals: &'a [Withdrawal],
+        store: Option<&'a Store>,
+        parent_hash: H256,
+    ) -> Self {
+        LevmProcessWithdrawalsIn {
+            block_cache,
+            withdrawals,
+            store,
+            parent_hash,
+        }
+    }
+}
 
-    for tx in block.body.transactions.iter() {
-        let report = execute_tx_levm(
-            tx,
-            block_header,
-            store_wrapper.clone(),
-            block_cache.clone(),
-            config,
-        )
-        .map_err(EvmError::from)?;
+#[derive(Debug)]
+pub struct LEVM;
 
-        let mut new_state = report.new_state.clone();
+impl IEVM for LEVM {
+    type Error = EvmError;
 
-        // Now original_value is going to be the same as the current_value, for the next transaction.
-        // It should have only one value but it is convenient to keep on using our CacheDB structure
-        for account in new_state.values_mut() {
-            for storage_slot in account.storage.values_mut() {
-                storage_slot.original_value = storage_slot.current_value;
+    type BlockExecutionOutput = (Vec<Receipt>, Vec<AccountUpdate>);
+
+    type TransactionExecutionInput<'a> = LevmTransactionExecutionIn<'a>;
+
+    type TransactionExecutionResult = ExecutionReport;
+
+    type GetStateTransitionsInput<'a> = LevmGetStateTransitionsIn<'a>;
+
+    type ProcessWithdrawalsInput<'a> = LevmProcessWithdrawalsIn<'a>;
+
+    fn execute_block(
+        block: &Block,
+        state: &mut EvmState,
+    ) -> Result<Self::BlockExecutionOutput, Self::Error> {
+        let store_wrapper = Arc::new(StoreWrapper {
+            store: state.database().unwrap().clone(),
+            block_hash: block.header.parent_hash,
+        });
+
+        let mut block_cache: CacheDB = HashMap::new();
+        let block_header = &block.header;
+        let config = state.chain_config()?;
+        cfg_if::cfg_if! {
+            if #[cfg(not(feature = "l2"))] {
```

Comment:
I believe we can just use the standard `cfg!` macro here:
```suggestion
        if cfg!(not(feature = "l2")) {
```
This has a few benefits:
1. Fewer proc_macro evaluations mean faster compile time (this would be a macro_rules macro instead);
2. I'm not sure how cfg_if behaves here, but using `cfg!` actually emits the code, but because the `if` evaluates a compile-time constant, dead-code removal leads to the same code, but we get type-check for all builds.